Content.Shared/_RMC14/Weapons/Ranged/CMGunSystem.cs

```diff
@@ -356,4 +364,44 @@ private void OnRevolverUniqueAction(Entity<RevolverAmmoProviderComponent> gun, r
 
         Dirty(gun);
     }
+
+    private void OnAmmoEjectActivateInWorld(Entity<RMCAmmoEjectComponent> gun, ref ActivateInWorldEvent args)
+    {
+        if (args.Handled ||
+            !_container.TryGetContainer(gun.Owner, gun.Comp.ContainerID, out var container) ||
+            container.ContainedEntities.Count <= 0 ||
+            !_hands.TryGetActiveHand(args.User, out var hand) ||
+            !hand.IsEmpty ||
+            !_hands.CanPickupToHand(args.User, container.ContainedEntities[0], hand))
+        {
+            return;
+        }
+
+        args.Handled = true;
+
+        var ejectedAmmo = container.ContainedEntities[0];
+
+        // For guns with a BallisticAmmoProviderComponent, if you just remove the ammo from its container, the gun system thinks it's still in the gun and you can still shoot it.
+        // So instead I'm having to inflict this shit on our codebase.
+        if (TryComp(gun.Owner, out BallisticAmmoProviderComponent? ammoProviderComponent))
+        {
+            var takeAmmoEvent = new TakeAmmoEvent(1, new List<(EntityUid?, IShootable)>(), Transform(gun.Owner).Coordinates, args.User);
+            RaiseLocalEvent(gun.Owner, takeAmmoEvent);
+
```

Comment:
Do you mean unspawnedcount or the contained entity or something else? If you're pulling it manually you can just call updateammocount manually iirc.

Reply:
I mean the `Entities` datafield in BallisticAmmoProviderComponent. It contains the gun's ammo and is not updated if I just remove the ammo from its container. Which makes it so if I just remove the rounds from their container, the gun can still fire them even when they're lying on the floor.

Raising a TakeAmmoEvent is the only way I found of removing ammo from that list, so I can actually pull it out of the gun properly.

UpdateAmmoCount is protected. It also, as far as I can tell, only affects the ammo counter UI.

Reply:
time to add access and fix it cheers on the bug report.